File: ios/TwitterFlux/Sources/Bridge/FluxBridge.swift
// FluxBridge — Swift wrapper around Flux C FFI.
// Uses @_silgen_name to declare C functions — no bridging header needed.

import Foundation
import Combine

// MARK: - C FFI declarations (linked from Rust static library)

@_silgen_name("flux_create")
private func _flux_create() -> OpaquePointer?

@_silgen_name("flux_free")
private func _flux_free(_ handle: OpaquePointer?)

@_silgen_name("flux_get")
private func _flux_get(_ handle: OpaquePointer?, _ path: UnsafePointer<CChar>?) -> _FluxBytes

@_silgen_name("flux_bytes_free")
private func _flux_bytes_free(_ bytes: _FluxBytes)

@_silgen_name("flux_emit")
private func _flux_emit(_ handle: OpaquePointer?, _ path: UnsafePointer<CChar>?, _ payload: UnsafePointer<CChar>?)

@_silgen_name("flux_server_url")
private func _flux_server_url(_ handle: OpaquePointer?) -> UnsafePointer<CChar>?

@_silgen_name("flux_i18n_get")
private func _flux_i18n_get(_ handle: OpaquePointer?, _ url: UnsafePointer<CChar>?) -> _FluxBytes

@_silgen_name("flux_i18n_set_locale")
private func _flux_i18n_set_locale(_ handle: OpaquePointer?, _ locale: UnsafePointer<CChar>?)

private struct _FluxBytes {
    let ptr: UnsafePointer<UInt8>?
    let len: Int
}

// MARK: - FluxStore (ObservableObject for SwiftUI)

/// Main interface to the Flux state engine from Swift.
///
/// Usage:
/// ```
/// @StateObject var store = FluxStore()
///
/// // Read state:
/// let auth: AuthState? = store.get("auth/state")
///
/// // Send request:
/// store.emit("auth/login", json: ["username": "alice"])
/// ```
final class FluxStore: ObservableObject {

    private let handle: OpaquePointer

    /// Bumped on every emit — SwiftUI observes this to trigger re-renders.
    @Published var revision: UInt64 = 0

    init() {
        guard let h = _flux_create() else {
            fatalError("flux_create() returned null")
        }
        handle = h
    }

    deinit {
        _flux_free(handle)
    }

    // MARK: - I18n

    /// Get a translated string synchronously.
    func t(_ url: String) -> String {
        let bytes = url.withCString { _flux_i18n_get(handle, $0) }
        defer { _flux_bytes_free(bytes) }
        guard bytes.len > 0, let ptr = bytes.ptr else { return url }
        return String(bytes: UnsafeBufferPointer(start: ptr, count: bytes.len), encoding: .utf8) ?? url
    }

    /// Set the i18n locale (e.g. "zh-CN", "en", "ja", "es").
    func setLocale(_ locale: String) {
        locale.withCString { _flux_i18n_set_locale(handle, $0) }
        objectWillChange.send()
        revision &+= 1
    }

    // MARK: - State

    /// Get state at a path, decoded as T.
    /// Call from SwiftUI views — reads `revision` to establish dependency.
    func get<T: Decodable>(_ path: String) -> T? {
        // Touch revision so SwiftUI knows to re-call when it changes.
        let _ = revision
        return getSync(path)
    }

    /// Get state without revision tracking (for non-SwiftUI code / tests).
    func getSync<T: Decodable>(_ path: String) -> T? {
        let bytes = path.withCString { _flux_get(handle, $0) }
        defer { _flux_bytes_free(bytes) }
        guard let ptr = bytes.ptr, bytes.len > 0 else { return nil }
        let data = Data(bytes: ptr, count: bytes.len)
        return try? JSONDecoder().decode(T.self, from: data)
    }

    // MARK: - Requests

    /// Emit a request with a JSON dictionary payload.
    func emit(_ path: String, json: [String: Any]) {
        if let data = try? JSONSerialization.data(withJSONObject: json),
           let str = String(data: data, encoding: .utf8) {
            path.withCString { p in
                str.withCString { j in
                    _flux_emit(handle, p, j)
                }
            }
        }
        objectWillChange.send()
        revision &+= 1
    }

    /// Emit a parameterless request.
    func emit(_ path: String) {
        path.withCString { _flux_emit(handle, $0, nil) }
        objectWillChange.send()
        revision &+= 1
    }

    // MARK: - Server Info

    /// The backend server URL (accessible from LAN).
    var serverURL: String {
        guard let ptr = _flux_server_url(handle) else { return "" }
        return String(cString: ptr)
    }

    /// Dashboard URL for the admin panel.
    var dashboardURL: URL? {
        URL(string: "\(serverURL)/dashboard")
    }
}

File: e2e/flux/flutter/lib/store/flux_bridge.dart
/// FluxBridge — Dart FFI wrapper around Flux C API.
///
/// Mirrors Swift's FluxBridge.swift exactly:
/// - flux_create / flux_free — lifecycle
/// - flux_get / flux_bytes_free — read state (JSON)
/// - flux_emit — dispatch request
/// - flux_i18n_get / flux_i18n_set_locale — translation
/// - flux_server_url — embedded server URL
library;

import 'dart:convert';
import 'dart:ffi';

import 'package:ffi/ffi.dart';

// ---------------------------------------------------------------------------
// C struct: FluxBytes { const uint8_t *ptr; size_t len; }
// ---------------------------------------------------------------------------

final class FluxBytes extends Struct {
  external Pointer<Uint8> ptr;

  @Size()
  external int len;
}

// ---------------------------------------------------------------------------
// FluxBridge
// ---------------------------------------------------------------------------

class FluxBridge {
  final DynamicLibrary _lib;
  final Pointer<Void> _handle;

  // Cached function pointers.
  late final FluxBytes Function(Pointer<Void>, Pointer<Utf8>) _getBytes;
  late final void Function(FluxBytes) _bytesFree;
  late final void Function(Pointer<Void>, Pointer<Utf8>, Pointer<Utf8>) _emit;
  late final FluxBytes Function(Pointer<Void>, Pointer<Utf8>) _i18nGet;
  late final void Function(Pointer<Void>, Pointer<Utf8>) _i18nSetLocale;
  late final Pointer<Utf8> Function(Pointer<Void>) _serverUrl;

  FluxBridge._(this._lib, this._handle) {
    _getBytes = _lib
        .lookupFunction<
          FluxBytes Function(Pointer<Void>, Pointer<Utf8>),
          FluxBytes Function(Pointer<Void>, Pointer<Utf8>)
        >('flux_get');

    _bytesFree = _lib
        .lookupFunction<Void Function(FluxBytes), void Function(FluxBytes)>(
          'flux_bytes_free',
        );

    _emit = _lib
        .lookupFunction<
          Void Function(Pointer<Void>, Pointer<Utf8>, Pointer<Utf8>),
          void Function(Pointer<Void>, Pointer<Utf8>, Pointer<Utf8>)
        >('flux_emit');

    _i18nGet = _lib
        .lookupFunction<
          FluxBytes Function(Pointer<Void>, Pointer<Utf8>),
          FluxBytes Function(Pointer<Void>, Pointer<Utf8>)
        >('flux_i18n_get');

    _i18nSetLocale = _lib
        .lookupFunction<
          Void Function(Pointer<Void>, Pointer<Utf8>),
          void Function(Pointer<Void>, Pointer<Utf8>)
        >('flux_i18n_set_locale');

    _serverUrl = _lib
        .lookupFunction<
          Pointer<Utf8> Function(Pointer<Void>),
          Pointer<Utf8> Function(Pointer<Void>)
        >('flux_server_url');
  }

  /// Open the Flux shared library at [libraryPath] and create a new engine.
  factory FluxBridge.open(String libraryPath) {
    final lib = DynamicLibrary.open(libraryPath);
    final create = lib
        .lookupFunction<Pointer<Void> Function(), Pointer<Void> Function()>(
          'flux_create',
        );

    final handle = create();
    if (handle == nullptr) {
      throw StateError('flux_create() returned null');
    }
    return FluxBridge._(lib, handle);
  }

  /// Free the Flux engine handle.
  void dispose() {
    final free = _lib
        .lookupFunction<
          Void Function(Pointer<Void>),
          void Function(Pointer<Void>)
        >('flux_free');
    free(_handle);
  }

  // -- State ---------------------------------------------------------------

  /// Read state at [path] as a JSON string. Returns `null` if not found.
  String? getJson(String path) {
    final pathPtr = path.toNativeUtf8();
    try {
      final bytes = _getBytes(_handle, pathPtr);
      if (bytes.ptr == nullptr || bytes.len == 0) return null;
      try {
        return utf8.decode(bytes.ptr.asTypedList(bytes.len));
      } finally {
        _bytesFree(bytes);
      }
    } finally {
      malloc.free(pathPtr);
    }
  }

  // -- Requests ------------------------------------------------------------

  /// Emit a request. [payloadJson] is optional (null for parameterless).
  void emit(String path, [String? payloadJson]) {
    final pathPtr = path.toNativeUtf8();
    final Pointer<Utf8> payloadPtr = payloadJson != null
        ? payloadJson.toNativeUtf8()
        : nullptr.cast<Utf8>();
    try {
      _emit(_handle, pathPtr, payloadPtr);
    } finally {
      malloc.free(pathPtr);
      if (payloadJson != null) malloc.free(payloadPtr);
    }
  }

  // -- I18n ----------------------------------------------------------------

  /// Get a translated string. Synchronous.
  /// [url] can be `"path"` or `"path?key=value&key2=value2"`.
  String t(String url) {
    final urlPtr = url.toNativeUtf8();
    try {
      final bytes = _i18nGet(_handle, urlPtr);
      if (bytes.ptr == nullptr || bytes.len == 0) return url;
      try {
        return utf8.decode(bytes.ptr.asTypedList(bytes.len));
      } finally {
        _bytesFree(bytes);
      }
    } finally {
      malloc.free(urlPtr);
    }
  }

  /// Set the i18n locale (e.g. "zh-CN", "en", "ja", "es").
  void setLocale(String locale) {
    final localePtr = locale.toNativeUtf8();
    try {
      _i18nSetLocale(_handle, localePtr);
    } finally {
      malloc.free(localePtr);
    }
  }

  // -- Server info ---------------------------------------------------------

  /// The embedded backend server URL (e.g. "http://192.168.1.100:3000").
  String get serverURL {
    final ptr = _serverUrl(_handle);
    if (ptr == nullptr) return '';
    return ptr.toDartString();
  }
}
